r1cs/include/r1cs/Dialect/IR/Ops.td

```diff
@@ -0,0 +1,124 @@
+//===- Ops.td ------------------------------------------------------------===//
+
+#ifndef R1CS_OPS
+#define R1CS_OPS
+include "mlir/IR/OpBase.td"
+include "mlir/IR/SymbolInterfaces.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "r1cs/Dialect/IR/Types.td"
+include "r1cs/Dialect/IR/Attrs.td"
+
+
+class R1CS_Op<string mnemonic, list<Trait> traits = []> :
+    Op<R1CSDialect, mnemonic, traits>;
+
+def R1CS_CircuitOp : R1CS_Op<"circuit", [
+  Symbol, IsolatedFromAbove, SingleBlock, NoTerminator
+]> {
+  let summary = "Circuit operation with named inputs and outputs";
+  let description = [{
+    This operation describes an R1CS circuit which specifies input, output, and intermediate signals
+    and consists of R1CS constraints. 
+    Example:
+
+    ```r1cs
+      r1cs.circuit @example inputs (%a: !r1cs.signal, %b: !r1cs.signal, %c: !r1cs.signal) {
+        %d = r1cs.def 1 : !r1cs.signal
+        %a_l = r1cs.to_linear %a : !r1cs.signal to !r1cs.linear
+        %b_l = r1cs.to_linear %b : !r1cs.signal to !r1cs.linear
+        !r1cs.constrain %d, %b, %a : !r1cs.linear
+        r1cs.return %d, %a : !r1cs.signal, !r1cs.signal
+      }
+    ```
+  }];
+
+  let arguments = (ins SymbolNameAttr:$sym_name);
+  let regions = (region SizedRegion<1>:$body);
+
+  let hasCustomAssemblyFormat = 1;
+
+  let builders = [
+    OpBuilder<(ins "::llvm::StringRef":$name, "::mlir::TypeRange":$resultTypes)>
+  ];
+
+  let extraClassDeclaration = [{
+    mlir::Block* addEntryBlock();
+  }];
+}
+
+def R1CS_ReturnOp : R1CS_Op<"return", [Terminator]> {
+  let summary = "Specifies output signals in circuit";
+  let arguments = (ins Variadic<R1CS_SignalType>:$results);
+  let description = [{
+      The `return` op returns a list of signal values that are supposed to be the output signals
+      of the circuit.
+      
+    }];
+
+  let assemblyFormat = [{
+    $results `:` type($results) attr-dict
+  }];
+}
+
+def SignalDefOp : R1CS_Op<"def", [Pure]> {
+  let summary = "Define a signal and its label ID";
+  let description = [{
+      The `def` op creates a fresh signal. Unless the signal appears in the return statement, it is considered
+      private and intermediate.
+  }];
+  let arguments = (ins UI64Attr:$label);
+  let results = (outs R1CS_SignalType:$out);
+  let assemblyFormat = [{
+    $label `:` type($out) attr-dict
```

Comment:
I think you could leave out the `type(x)` from most of the ops since they are a specific simple type. See the Buildable Types section of https://mlir.llvm.org/docs/DefiningDialects/Operations/#type-inference